shared-bindings/network/__init__.c

```diff
@@ -0,0 +1,92 @@
+ *
+ * Copyright (c) 2014 Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "py/objlist.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "lib/netutils/netutils.h"
+
+#include "shared-bindings/network/__init__.h"
+
+#if MICROPY_PY_NETWORK
+
+/// \module network - network configuration
+///
+/// This module provides network drivers and routing configuration.
+
+void network_module_init(void) {
+    mp_obj_list_init(&MP_STATE_PORT(mod_network_nic_list), 0);
+}
+
+void network_module_deinit(void) {
+}
+
+void network_module_register_nic(mp_obj_t nic) {
+    for (mp_uint_t i = 0; i < MP_STATE_PORT(mod_network_nic_list).len; i++) {
+        if (MP_STATE_PORT(mod_network_nic_list).items[i] == nic) {
+            // nic already registered
+            return;
+        }
+    }
+    // nic not registered so add to list
+    mp_obj_list_append(MP_OBJ_FROM_PTR(&MP_STATE_PORT(mod_network_nic_list)), nic);
+}
+
+mp_obj_t network_module_find_nic(const uint8_t *ip) {
+    // find a NIC that is suited to given IP address
+    for (mp_uint_t i = 0; i < MP_STATE_PORT(mod_network_nic_list).len; i++) {
+        mp_obj_t nic = MP_STATE_PORT(mod_network_nic_list).items[i];
+        // TODO check IP suitability here
+        //mod_network_nic_type_t *nic_type = (mod_network_nic_type_t*)mp_obj_get_type(nic);
+        return nic;
+    }
+
+    nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, translate("no available NIC")));
+}
+
+STATIC mp_obj_t network_route(void) {
+    return MP_OBJ_FROM_PTR(&MP_STATE_PORT(mod_network_nic_list));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(network_route_obj, network_route);
+
+STATIC const mp_rom_map_elem_t mp_module_network_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_network) },
+    { MP_ROM_QSTR(MP_QSTR_route), MP_ROM_PTR(&network_route_obj) },
```

Comment:
What is this useful for? Can we omit the network module for now?

Reply:
The network module maintains a registry of nics, and when you configure a nic (eg: construct it with `wiznet.WIZNET5K()`) it calls `network_module_register_nic` to add that nic to the list.  When you call `socket.socket()` it calls `network_module_find_nic` to find an appropriate nic for this socket.

This functionality could be merged into the socket module I suppose.

Reply:
... in MicroPython, the network module also collects all the network device classes in one place (they're not all compiled in, you can control that with various MICROPY_PY_WHATEVER flags)

I've split WIZNET5K out into its own 'wiznet' module, but I'm not sure that once we have a few of these we wouldn't be better collecting them into 'network' after all ...

Reply:
I think network is the way to go. I don't want extra stuff in socket.

Is it useful from Python though? Maybe it should be an internal only concept for now.

Reply:
Yeah, in MicroPython it exposes the nic classes.  In CircuitPython all it has left is 'network.route' which lets you enumerate all the configured nics, eg:
```
>>> network.route()[0].ifconfig()
('10.107.1.123', '255.255.255.0', '10.107.1.1', '0.0.0.0')
```
not necessarily very useful

Reply:
How about making it internal only for now?